File: Admin Client/e_jam_backend/src/models/processes.rs
use serde::{Deserialize, Serialize};

#[doc = r"Processes that are running on the device either a verification process or a generation process
each process has a status (Type: ProcessStatus) that can be one of the following
Idle, Running, Offline, Paused, Completed, Failed
each process has a name that is a string that represents both the name of the process and the device that is running the process (Device IP address + Port number)
## Variants
* `Generation` (the process is a generation process)
* `Verification` (the process is a verification process)
* `GenerationAndVerification` (the process is a generation and verification process)"]
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(tag = "processtype")]
pub enum ProcessType {
    #[serde(rename = "Generation")]
    Generation,
    #[serde(rename = "Verification")]
    Verification,
    #[serde(rename = "GenerationAndVerification")]
    GenerationaAndVerification,
}

#[doc = r"process status
a process has a status that represents the current state of the stream in the specific device
each process has a type (Type: ProcessType) that can be one of the following
Generation, Verification, GenerationAndVerification
each process has a name that is a string that represents both the name of the process and the device that is running the process (Device IP address + Port number)
## Variants
* `Queued` (the process is idle and waiting to be started)
* `Running` (the process is running) 
* `Offline` (the process is offline)
* `Paused` (the process is paused)
* `Completed` (the process is successfully completed)
* `Failed` (the process is failed)"]
#[derive(Serialize, Deserialize, Default, Debug, Clone, PartialEq)]
#[serde(tag = "processstatus")]
pub enum ProcessStatus {
    #[default]
    #[serde(rename = "Idle")]
    // the process is idle
    Queued,

    #[serde(rename = "Running")]
    // the process is running
    Running,

    #[serde(rename = "Offline")]
    // the process is offline
    Offline,

    #[serde(rename = "Paused")]
    // the process is paused
    Paused,

    #[serde(rename = "Completed")]
    // the process is successfully completed
    Completed,

    #[serde(rename = "Failed")]
    // the process is failed
    Failed,
}

File: Admin Client/e_jam_frontend/lib/src/Model/Enums/processes.dart
enum ProcessStatus {
  queued,
  running,
  stopped,
  completed,
  failed,
}

ProcessStatus processStatusFromString(String status) {
  switch (status) {
    case 'Queued':
      return ProcessStatus.queued;
    case 'Running':
      return ProcessStatus.running;
    case 'Stopped':
      return ProcessStatus.stopped;
    case 'Completed':
      return ProcessStatus.completed;
    case 'Failed':
      return ProcessStatus.failed;
    default:
      return ProcessStatus.failed;
  }
}

String processStatusToString(ProcessStatus? processStatus) {
  switch (processStatus) {
    case ProcessStatus.queued:
      return 'Queued';
    case ProcessStatus.running:
      return 'Running';
    case ProcessStatus.stopped:
      return 'Stopped';
    case ProcessStatus.completed:
      return 'Completed';
    case ProcessStatus.failed:
      return 'Failed';
    default:
      return 'Failed';
  }
}

enum ProcessType {
  generation,
  verification,
  generatingAndVerification,
}

ProcessType processTypeFromString(String type) {
  switch (type) {
    case 'Generation':
      return ProcessType.generation;
    case 'Verification':
      return ProcessType.verification;
    case 'GeneratingAndVerification':
      return ProcessType.generatingAndVerification;
    default:
      return ProcessType.generatingAndVerification;
  }
}

String processTypeToString(ProcessType processType) {
  switch (processType) {
    case ProcessType.generation:
      return 'Generation';
    case ProcessType.verification:
      return 'Verification';
    case ProcessType.generatingAndVerification:
      return 'GeneratingAndVerification';
    default:
      return 'GeneratingAndVerification';
  }
}
